src/metadata.rs

```diff
@@ -25,30 +25,165 @@ pub struct Metadata {
+
+        let year: Option<i32> = check_pattern!(
+            PATTERN_YEAR,
+            name,
+            title_start,
+            title_end,
+            |caps: Captures<'_>| caps
+                .name("year")
+                .map(|m| m.as_str())
+                .map(|s| s.parse().unwrap())
+        );
+
+        let resolution = check_pattern!(
+            PATTERN_RESOLUTION,
+            name,
+            title_start,
+            title_end,
+            |caps: Captures<'_>| caps.get(1).map(|m| m.as_str().to_string())
+        );
+        let quality = check_pattern!(
+            PATTERN_QUALITY,
+            name,
+            title_start,
+            title_end,
+            |caps: Captures<'_>| caps.get(0).map(|m| m.as_str().to_string())
+        );
+        let codec = check_pattern!(
+            PATTERN_CODEC,
+            name,
+            title_start,
+            title_end,
+            |caps: Captures<'_>| caps.get(0).map(|m| m.as_str().to_string())
+        );
+        let audio = check_pattern!(
+            PATTERN_AUDIO,
+            name,
+            title_start,
+            title_end,
+            |caps: Captures<'_>| caps.get(0).map(|m| m.as_str().to_string())
+        );
+        let group = check_pattern!(
+            PATTERN_GROUP,
+            name,
+            title_start,
+            title_end,
+            |caps: Captures<'_>| caps.get(2).map(|m| m.as_str().to_string())
+        );
+        let imdb = check_pattern!(
+            PATTERN_IMDB,
+            name,
+            title_start,
+            title_end,
+            |caps: Captures<'_>| caps.get(0).map(|m| m.as_str().to_string())
+        );
+
+        let extended = check_pattern!(PATTERN_EXTENDED, name, title_start, title_end);
+        let hardcoded = check_pattern!(PATTERN_HARDCODED, name, title_start, title_end);
+        let proper = check_pattern!(PATTERN_PROPER, name, title_start, title_end);
+        let repack = check_pattern!(PATTERN_REPACK, name, title_start, title_end);
+        let widescreen = check_pattern!(PATTERN_WIDESCREEN, name, title_start, title_end);
+        let unrated = check_pattern!(PATTERN_UNRATED, name, title_start, title_end);
+        let three_d = check_pattern!(PATTERN_THREE_D, name, title_start, title_end);
+
+        let region = check_pattern!(PATTERN_REGION, name, title_start, title_end);
+        let container = check_pattern!(PATTERN_CONTAINER, name, title_start, title_end);
+        let language = check_pattern!(PATTERN_LANGUAGE, name, title_start, title_end);
+        let garbage = check_pattern!(PATTERN_GARBAGE, name, title_start, title_end);
+        let website = check_pattern!(PATTERN_WEBSITE, name, title_start, title_end);
 
         if title_start >= title_end {
-            return Err(ErrorMatch::new(
-                captures.into_iter().map(|(k, v)| (k, v.get(0))).collect(),
-            ));
+            return Err(ErrorMatch::new(vec![
+                ("season", season.map(|s| s.to_string())),
```

Comment:
just `.map(String::from)`

Reply:
This doesn't work with integer primitives (they `impl ToString` but there's no `From` impl on `String` for them), but I changed it to defer parsing until after the error check anyway (bonus: only allocate if there is an error), so this works